src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,171 @@
+			solMapList.add(e);
+			map.put(valKeys, solMapList);
+		}
+		return true;
 	}
 
 	@Override
 	public void clear() {
-		// TODO Auto-generated method stub
-		
+		map.clear();
 	}
 
 	@Override
 	public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		final List<Node> valKeys = getVarKeys(sm);
+		final List<SolutionMapping> joinPartner = new ArrayList<>();
+		List<SolutionMapping> matchingSolMaps = new ArrayList<>();
+
+		if (!valKeys.isEmpty() && valKeys.size() == joinVariables.size()){
+			matchingSolMaps = map.get(valKeys);
+		}
+		else {
+			final Iterator<List<SolutionMapping>> li = map.values().iterator();
+			while(li.hasNext()){
+				matchingSolMaps.addAll(li.next());
+			}
+		}
+
+		if (matchingSolMaps != null) {
+			for (SolutionMapping matchSolM : matchingSolMaps) {
+				if (SolutionMappingUtils.compatible(sm, matchSolM)) {
+					joinPartner.add(matchSolM);
+				}
+			}
+		}
+		return joinPartner;
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings( final Var var, final Node value )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 1 && joinVariables.contains(var) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value);
+			final List<SolutionMapping> solMaps = map.get(valKeyL);
+			return solMaps;
+		}
+		else{
+			throw new IllegalArgumentException();
+		}
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings(
 			final Var var1, final Node value1,
 			final Var var2, final Node value2 )
-					throws UnsupportedOperationException
+			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 2 && joinVariables.contains(var1) && joinVariables.contains(var2) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value1);
+			valKeyL.add(value2);
+			final List<SolutionMapping> solMaps = map.get(valKeyL);
+			return solMaps;
+		}
+		else{
+			throw new IllegalArgumentException();
+		}
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings(
 			final Var var1, final Node value1,
 			final Var var2, final Node value2,
 			final Var var3, final Node value3 )
-					throws UnsupportedOperationException
+			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if( joinVariables.size() == 3 && joinVariables.contains(var1) && joinVariables.contains(var2) && joinVariables.contains(var3) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value1);
+			valKeyL.add(value2);
+			valKeyL.add(value3);
+			final List<SolutionMapping> solMaps = map.get(valKeyL);
+			return solMaps;
+		}
+		else{
+			throw new IllegalArgumentException();
+		}
 	}
 
+	public List<Node> getVarKeys(final SolutionMapping e){
```

Comment:
```suggestion
	protected List<Node> getVarKeys(final SolutionMapping e){
```

Reply:
In addition to the suggested change, I suggest to rename this method to `getVarKey` because it produces only one index key.